Question1: Predicting Heart Disease

In this question, we work with a dataset from the great textbook of "An Introduction to Statistical Learning."

a- Read the dataset file “Hearts_s.csv” (you should download it from CSNS), and assign it to a Pandas DataFrame.

In [1]:

import numpy as np
import pandas as pd

heart_df = pd.read_csv('C:/RiddhiDocument/MS Computer Science/CS-4661 Data Sceince/Homework/HW3/Heart_s.csv')


print(heart_df)

     Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
5     56      f    nontypical     120   236        0    178      0.8   
6     62      m  asymptomatic     140   268        2    160      3.6   
7     57      m  asymptomatic     120   354        0    163      0.6   
8     63      f  asymptomatic     130   254        2    147      1.4   
9     53      f  asymptomatic     140   203        2    155      3.1   
10    57      f  asymptomatic     140   192        0    148      0.4   
11    56      m    nontypical     140   294        2    153      1.3   
12    56      f    nonanginal     130   256        2    142     

In [2]:
heart_df

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
5     56      f    nontypical     120   236        0    178      0.8   
6     62      m  asymptomatic     140   268        2    160      3.6   
7     57      m  asymptomatic     120   354        0    163      0.6   
8     63      f  asymptomatic     130   254        2    147      1.4   
9     53      f  asymptomatic     140   203        2    155      3.1   
10    57      f  asymptomatic     140   192        0    148      0.4   
11    56      m    nontypical     140   294        2    153      1.3   
12    56      f    nonanginal     130   256        2    142      0.6   
13    44      f    nontypical     120   263        0    173      0.0   
14    52      f    nonanginal     172   199        0    162      0.5   
15    57      f    nonanginal     150   168        0    174      1.6   
16    48      f    nontypical     110   229        0    168      1.0   
17    54      f  asymptomatic     140   239        0    160      1.2   
18    48      m    nonanginal     130   275        0    139      0.2   
19    49      f    nontypical     130   266        0    171      0.6   
20    64      f       typical     110   211        2    144      1.8   
21    58      m       typical     150   283        2    162      1.0   
22    58      f    nontypical     120   284        2    160      1.8   
23    58      f    nonanginal     132   224        2    173      3.2   
24    60      f  asymptomatic     130   206        2    132      2.4   
25    50      m    nonanginal     120   219        0    158      1.6   
26    58      m    nonanginal     120   340        0    172      0.0   
27    66      m       typical     150   226        0    114      2.6   
28    43      f  asymptomatic     150   247        0    171      1.5   
29    40      f  asymptomatic     110   167        2    114      2.0   
..   ...    ...           ...     ...   ...      ...    ...      ...   
271   71      m  asymptomatic     112   149        0    125      1.6   
272   59      f       typical     134   204        0    162      0.8   
273   64      f       typical     170   227        2    155      0.6   
274   66      m    nonanginal     146   278        2    152      0.0   
275   39      m    nonanginal     138   220        0    152      0.0   
276   57      f    nontypical     154   232        2    164      0.0   
277   58      m  asymptomatic     130   197        0    131      0.6   
278   57      f  asymptomatic     110   335        0    143      3.0   
279   47      f    nonanginal     130   253        0    179      0.0   
280   55      m  asymptomatic     128   205        1    130      2.0   
281   35      f    nontypical     122   192        0    174      0.0   
282   61      f  asymptomatic     148   203        0    161      0.0   
283   58      f  asymptomatic     114   318        1    140      4.4   
284   58      m  asymptomatic     170   225        2    146      2.8   
285   58      f    nontypical     125   220        0    144      0.4   
286   56      f    nontypical     130   221        2    163      0.0   
287   56      f    nontypical     120   240        0    169      0.0   
288   67      f    nonanginal     152   212        2    150      0.8   
289   55      m    nontypical     132   342        0    166      1.2   
290   44      f  asymptomatic     120   169        0    144      2.8   
291   63      f  asymptomatic     140   187        2    144      4.0   
292   63      m  asymptomatic     124   197        0    136      0.0   
293   41      f    nontypical     120   157        0    182      0.0   
294   59      f  asymptomatic     164   176  

b- Check out the dataset. As you see, the dataset contains a number of features including both contextual and biological 
factors (e.g. age, gender, vital signs, …). The last column “AHD” is the label with “Yes” meaning that a human subject has 
Heart Disease, and “No” meaning that the subject does not have Heart Disease.

In [3]:
# checking the dataset by printing every 10 lines:
heart_df[0::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  
0         fixed   No  
10        fixed   No  
20       normal   No  
30       normal   No  
40   reversable  Yes  
50       normal   No  
60   reversable  Yes  
70       normal   No  
80       normal   No  
90   reversable  Yes  
100      normal   No  
110      normal  Yes  
120  reversable  Yes  
130  reversable   No  
140  reversable  Yes  
150      normal   No  
160      normal  Yes  
170  reversable   No  
180  reversable  Yes  
190  reversable  Yes  
200      normal   No  
210  reversable  Yes  
220      normal   No  
230      normal  Yes  
240      normal   No  
250  reversable  Yes  
260      normal  Yes  
270  reversable  Yes  
280  reversable  Yes  
290       fixed  Yes  
300      normal   No

c- As you see, there are at least 3 categorical features in the dataset (Gender, ChestPain, Thal). Let’s ignore these 
categorical features for now, only keep the numerical features and build your feature matrix and label vector.


In [4]:
feat_cols = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak']

X = heart_df[feat_cols]

y = heart_df['AHD']

#print(X.head()) # print the first 5 rows
#print(y.head()) # print the first 5 rows

print(X)
print(y)

     Age  RestBP  Chol  RestECG  MaxHR  Oldpeak
0     63     145   233        2    150      2.3
1     67     160   286        2    108      1.5
2     67     120   229        2    129      2.6
3     37     130   250        0    187      3.5
4     41     130   204        2    172      1.4
5     56     120   236        0    178      0.8
6     62     140   268        2    160      3.6
7     57     120   354        0    163      0.6
8     63     130   254        2    147      1.4
9     53     140   203        2    155      3.1
10    57     140   192        0    148      0.4
11    56     140   294        2    153      1.3
12    56     130   256        2    142      0.6
13    44     120   263        0    173      0.0
14    52     172   199        0    162      0.5
15    57     150   168        0    174      1.6
16    48     110   229        0    168      1.0
17    54     140   239        0    160      1.2
18    48     130   275        0    139      0.2
19    49     130   266        0    171  

d- Split the dataset into testing and training sets with the following parameters: test_size=0.25, random_state=4

In [5]:

from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=4)

print("X_train.shape :",X_train.shape)
print("y_train.shape :",y_train.shape)


print("X_test.shape :",X_test.shape)
print("y_test.shape :",y_test.shape)


X_train.shape : (225, 6)
y_train.shape : (225,)
X_test.shape : (76, 6)
y_test.shape : (76,)


e- Use KNN (with k=3), Decision Tree (with random_state=5), and Logistic Regression Classifiers to predict Heart Disease based 
on the training/testing datasets that you built in part (d). Then check, compare, and report the accuracy of these 3 
classifiers. Which one is the best? Which one is the worst?

In [6]:
from sklearn.neighbors import KNeighborsClassifier

k = 3

knn = KNeighborsClassifier(n_neighbors=k)

knn.fit(X_train,y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [7]:
y_predict = knn.predict(X_test)

print(y_predict)

['Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No'
 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No'
 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes'
 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes'
 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No']


In [8]:
from sklearn.metrics import accuracy_score

accu_knn = accuracy_score(y_test,y_predict)

print("KNN Accuracy : ",accu_knn)

KNN Accuracy :  0.6973684210526315


In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

logRegr = LogisticRegression()

deciTreeCla = DecisionTreeClassifier()

logRegr.fit(X_train,y_train)
deciTreeCla.fit(X_train,y_train)


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [10]:
y_predict_lr = logRegr.predict(X_test)

y_predict_dt = deciTreeCla.predict(X_test)

print("y_Predict Linear Regression : ",y_predict_lr)
print("y_predict Decission Tree : ",y_predict_dt)

y_Predict Linear Regression :  ['Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes'
 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No'
 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes'
 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes'
 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No'
 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No']
y_predict Decission Tree :  ['Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes'
 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes'
 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes'
 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'Yes'
 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes'
 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes']


In [11]:
accu_lr = accuracy_score(y_test,y_predict_lr)
accu_dt = accuracy_score(y_test,y_predict_dt)

print("Logistic Regression : ",accu_lr)
print("Decission Tree :", accu_dt)
print("KNN : ",accu_knn)

Logistic Regression :  0.7236842105263158
Decission Tree : 0.6578947368421053
KNN :  0.6973684210526315


  Which one is the best ? Which one si the worst?

ANS : Logistic Regression have best accuracy : 0.7236842105263158
      And Decision Tree classifier have worst accuracy : 0.5921052631578947

f- Now, we want to use the categorical features as well! To this end, we have to perform a feature engineering process called
OneHotEncoding for the categorical features. To do this, each categorical feature should be replaced with dummy columns in the feature
table (one column for each possible value of a categorical feature), and then encode it in a binary manner such that only one of the dummy columns can take “1” at a time (and
zero for the rest). For example, “Gender” can take two values “m” and “f”. Thus, we need to replace this feature 
(in the feature table) by 2 columns titled “m” and “f”. Wherever we have a male subject, we can put “1” and ”0” in the columns “m” and “f”. Wherever
we have a female subject, we can put “0” and ”1” in the columns “m” and “f”. (Hint: you
will need 4 columns to encode “ChestPain” and 3 columns to encode “Thal”).

In [12]:
gender = pd.get_dummies(heart_df['Gender'])
chestpain = pd.get_dummies(heart_df['ChestPain'])
thal = pd.get_dummies(heart_df['Thal'])

heart_df = heart_df.drop('Gender',axis=1)
heart_df = heart_df.drop('ChestPain',axis=1)
heart_df = heart_df.drop('Thal',axis=1)

heart_df = heart_df.join(gender)
heart_df = heart_df.join(chestpain)
heart_df = heart_df.join(thal)

heart_df

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  AHD  f  m  asymptomatic  \
0     63     145   233        2    150      2.3   No  1  0             0   
1     67     160   286        2    108      1.5  Yes  1  0             1   
2     67     120   229        2    129      2.6  Yes  1  0             1   
3     37     130   250        0    187      3.5   No  1  0             0   
4     41     130   204        2    172      1.4   No  0  1             0   
5     56     120   236        0    178      0.8   No  1  0             0   
6     62     140   268        2    160      3.6  Yes  0  1             1   
7     57     120   354        0    163      0.6   No  0  1             1   
8     63     130   254        2    147      1.4  Yes  1  0             1   
9     53     140   203        2    155      3.1  Yes  1  0             1   
10    57     140   192        0    148      0.4   No  1  0             1   
11    56     140   294        2    153      1.3   No  0  1             0   
12    56     130   256        2    142      0.6  Yes  1  0             0   
13    44     120   263        0    173      0.0   No  1  0             0   
14    52     172   199        0    162      0.5   No  1  0             0   
15    57     150   168        0    174      1.6   No  1  0             0   
16    48     110   229        0    168      1.0  Yes  1  0             0   
17    54     140   239        0    160      1.2   No  1  0             1   
18    48     130   275        0    139      0.2   No  0  1             0   
19    49     130   266        0    171      0.6   No  1  0             0   
20    64     110   211        2    144      1.8   No  1  0             0   
21    58     150   283        2    162      1.0   No  0  1             0   
22    58     120   284        2    160      1.8  Yes  1  0             0   
23    58     132   224        2    173      3.2  Yes  1  0             0   
24    60     130   206        2    132      2.4  Yes  1  0             1   
25    50     120   219        0    158      1.6   No  0  1             0   
26    58     120   340        0    172      0.0   No  0  1             0   
27    66     150   226        0    114      2.6   No  0  1             0   
28    43     150   247        0    171      1.5   No  1  0             1   
29    40     110   167        2    114      2.0  Yes  1  0             1   
..   ...     ...   ...      ...    ...      ...  ... .. ..           ...   
271   71     112   149        0    125      1.6   No  0  1             1   
272   59     134   204        0    162      0.8  Yes  1  0             0   
273   64     170   227        2    155      0.6   No  1  0             0   
274   66     146   278        2    152      0.0   No  0  1             0   
275   39     138   220        0    152      0.0   No  0  1             0   
276   57     154   232        2    164      0.0  Yes  1  0             0   
277   58     130   197        0    131      0.6   No  0  1             1   
278   57     110   335        0    143      3.0  Yes  1  0             1   
279   47     130   253        0    179      0.0   No  1  0             0   
280   55     128   205        1    130      2.0  Yes  0  1             1   
281   35     122   192        0    174      0.0   No  1  0             0   
282   61     148   203        0    161      0.0  Yes  1  0             1   
283   58     114   318        1    140      4.4  Yes  1  0             1   
284   58     170   225        2    146      2.8  Yes  0  1             1   
285   58     125   220        0    144      0.4   No  1  0             0   
286   56     130   221        2    163      0.0   No  1  0             0   
287   56     120   240        0    169      0.0   No  1  0             0   
288   67     152   212        2    150      0.8  Yes  1  0             0   
289   55     132   342        0    166      1.2   No  0  1             0   
290   44     120   169        0    144      2.8  Yes  1  0             1   
291   63     140   187        2    144      4.0  Yes 

g- Repeat parts (d) and (e) with the new dataset that you built in part (f). How does the prediction accuracy change for each method?


In [13]:
n = heart_df.columns.values

newfeatu = ['Age','RestBP','Chol','RestECG','MaxHR','Oldpeak','f','m','asymptomatic','nonanginal','nontypical','typical','fixed'
           ,'normal','reversable']

new_X = heart_df[newfeatu]
new_y = heart_df['AHD']

print(new_X.head())
print(new_y.head())

   Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  f  m  asymptomatic  nonanginal  \
0   63     145   233        2    150      2.3  1  0             0           0   
1   67     160   286        2    108      1.5  1  0             1           0   
2   67     120   229        2    129      2.6  1  0             1           0   
3   37     130   250        0    187      3.5  1  0             0           1   
4   41     130   204        2    172      1.4  0  1             0           0   

   nontypical  typical  fixed  normal  reversable  
0           0        1      1       0           0  
1           0        0      0       1           0  
2           0        0      0       0           1  
3           0        0      0       1           0  
4           1        0      0       1           0  
0     No
1    Yes
2    Yes
3     No
4     No
Name: AHD, dtype: object


In [14]:
x_gtrain,x_gtest,y_gtrain,y_gtest = train_test_split(new_X,new_y,test_size=0.25,random_state=5)
print(x_gtrain.head())

k =3
new_knn = KNeighborsClassifier(n_neighbors=k)

new_knn.fit(x_gtrain,y_gtrain)

     Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  f  m  asymptomatic  \
259   44     118   242        0    149      0.3  0  1             0   
289   55     132   342        0    166      1.2  0  1             0   
184   63     140   195        0    179      0.0  0  1             0   
37    57     150   276        2    112      0.6  1  0             1   
299   57     130   236        2    174      0.0  0  1             0   

     nonanginal  nontypical  typical  fixed  normal  reversable  
259           1           0        0      0       1           0  
289           0           1        0      0       1           0  
184           0           1        0      0       1           0  
37            0           0        0      1       0           0  
299           0           1        0      0       1           0  


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [15]:
y_gpredict = new_knn.predict(x_gtest)
print("y_gpredict : ",y_gpredict)

new_accuracy_knn = accuracy_score(y_gtest,y_gpredict)
print("\n\n KNN Accuracy after OneHotCodeing :",new_accuracy_knn)

y_gpredict :  ['No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No'
 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No'
 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes'
 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'Yes' 'No' 'No'
 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No']


 KNN Accuracy after OneHotCodeing : 0.7236842105263158


In [16]:
new_logregr = LogisticRegression()
new_decesionTreeCla = DecisionTreeClassifier()

new_logregr.fit(x_gtrain,y_gtrain)
new_decesionTreeCla.fit(x_gtrain,y_gtrain)


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [17]:
y_gpredict_lr = new_logregr.predict(x_gtest)
y_gpredict_dt = new_decesionTreeCla.predict(x_gtest)

print("\n y_gpredict_lr : ",y_gpredict_lr)
print("\n y_gpredict_dt : ",y_gpredict_dt)


 y_gpredict_lr :  ['Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No'
 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No'
 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No'
 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes'
 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'Yes']

 y_gpredict_dt :  ['Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No'
 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No'
 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes'
 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes'
 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No'
 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes']


In [18]:
new_accuracy_lr = accuracy_score(y_gtest,y_gpredict_lr)
new_accuracy_dt = accuracy_score(y_gtest,y_gpredict_dt)

print("Logistic Regression : ",new_accuracy_lr )
print("Decission Tree Classifier : ",new_accuracy_dt )
print("KNN Accuracy : ",new_accuracy_knn )

Logistic Regression :  0.8947368421052632
Decission Tree Classifier :  0.7763157894736842
KNN Accuracy :  0.7236842105263158


h- Now, repeat part (e) with the new dataset that you built in part (f), but this time using Cross-Validation. Thus, rather 
than splitting the dataset into testing and training, use 10-fold Cross-Validation (as we learned in Lab4) to evaluate the 
classification methods and report the final prediction accuracy.

In [19]:
from sklearn.cross_validation import cross_val_score

k=3

myknn = KNeighborsClassifier(n_neighbors=k)
mylogreg = LogisticRegression()
myDecisionTree = DecisionTreeClassifier()

knn_accu_list = cross_val_score(myknn,new_X,new_y,cv=10,scoring='accuracy')
lr_accu_list = cross_val_score(mylogreg,new_X,new_y,cv=10,scoring='accuracy')
dt_accu_list = cross_val_score(myDecisionTree,new_X,new_y,cv=10,scoring='accuracy')


print(knn_accu_list)
print(lr_accu_list)
print(dt_accu_list)

C:\Users\ashish\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


[0.70967742 0.64516129 0.51612903 0.66666667 0.63333333 0.46666667
 0.66666667 0.66666667 0.55172414 0.72413793]
[0.77419355 0.80645161 0.83870968 0.93333333 0.9        0.66666667
 0.8        0.83333333 0.79310345 0.79310345]
[0.74193548 0.67741935 0.77419355 0.73333333 0.76666667 0.6
 0.6        0.6        0.68965517 0.72413793]


In [20]:
knn_accuracy = knn_accu_list.mean()
lr_accuracy = lr_accu_list.mean()
dt_accuracy = dt_accu_list.mean()

print("KNN accuracy :",knn_accuracy)
print("Logitic Regression accuracy :",lr_accuracy)
print("Decision Tree Classifier accuracy :",dt_accuracy)


KNN accuracy : 0.6246829810901001
Logitic Regression accuracy : 0.8138895068594735
Decision Tree Classifier accuracy : 0.6907341490545049
